### DISCUSSION TITLE
Web interface

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#35] [Show and tell]
Hey folks,
I guess you noticed the new web interface is just a static chess board only,,, I have zero experience in web design. I only knew to deploy the server and build a basic web app. We now have the structure in there so hopefully we'll board someone with experience.

It's very easy though to bring that board alive. Inside the template file there is a FEN that control position so I need to build a way to pass that as a variable from stockfish pipe inside the template... For lichess I guess you can just push it from python-chess.

Other thing I am looking at now is to use flask-turbo to enable browser auto-refresh once board position chages.

**Comment by EdNekebno** [comment]
Yup, for lichess/eboard scripts etc we could probably push somehow (I'm not familiar with flask). I see you put a project thing for SQLite. Would it make sense that we modify each script to fill in a database and stockfish to fill it in for the centaur process, and then use the database as the source for displaying this board in the web app and for PGN generation?   (in the land of faraway future dreams - that may enable the web interface in the future to do some game analysis too).

**Comment by wormstein** [comment]
Me neither. No experience at all...
From what I read calls go in app.py once the user access the webpage. In there it is python... Then flask calls templates rendering toghether with the variables and renders the webpage.

For stockfish, the one I have on board, writes the FEN in realtime in a file that could be read by python and pass to fask, For other scripts we could do that way: store info in a database from where we can read it.

I think SQLite will be a powerfull tool for us. Thing like settings, game history, fens, etc. can be stored in there. Is just I want to have it with SQLAlchemy.

Cheers

**Comment by EdNekebno** [comment]
Once we know the table structure, I can probably adapt the stockfish adaption to go straight into SQLite if you like. If there's some sort of source field then you can have (stockfish, lichess, eboard, etc) and everything works the same way. Then you build a script that can convert the current game from that to a FEN and pass that through to the web interface and it's impartial to what process is actually playing the game at the time. Being a python newb, I've not heard of SQLAlchemy. Looks good.

**Comment by EdNekebno** [comment]
I have taken some liberties with your web interface :) Start the new menu.py and visit the web interface. Scroll up down (on dgt centaur menu).

The integration is only by a very basic method. But I think it starts to show how the new idea for the core functionality of boardfunctions can interact with the web interface.

It works in quite a klunky way at the moment. I think what is needed is probably websockets so that boardfunctions.py can message the frontend of the web app to get it to do move pieces, refresh screen, etc. This can also work in reverse - if the board pieces are out of sync in a game of the board the web interface can allow correction, it would be possible to make the menu clickable on the web interface to choose functions on the actual board, even the buttons could work if we put them on there - in everything except the actual official 'centaur' process [although even then we can potentially inject into that process].

By the way - I could only get flask to show changes after restarting the whole service. It is kind of slow to develop then. I wonder if there is a quicker way or if we'll have to live with it.

**Comment by wormstein** [comment]
I am tring to imagine what the code does as I'm not at home to give it a go :))
Ia it showing what is on display to the web interface?

Side note, I am almost ready to push the actual change to animate the board... It's actually vrey easy, as I just need to write a function in the route / block inside app.py and pass fen once render_template is returned,

The line in the template is already there to grab the FEN:
`var ruyLopez = '{{ fen }}`

For web interface interaction, I already planned to use IO Socket fot bidirectional for comm changes, This wsy we will be able to pass movements from browser to the board and back. We could even have someone play in browser and the board lights up the fields once a move is made...
Chessboard.js is very similar to python-chess in functionalities.

**Comment by EdNekebno** [comment]
The menu uses the new epaper display code.
The new epaper display code runs as a thread and essentially monitors a Pillow image for changes and sends it to the epaper. So it also drops it into the web static folder.
At the moment I made the web interface just reload this image every second (as epaper refreshes 1.2 secs). But when we have bi-directional communication the epaper thread can theoretically tell the web interface it has updated and it can just update then.

The new functions I worked on in boardfunctions.py to pass keyPresses, fieldActivity, and more also run threaded. So when there is bi-directional communication then it could send this data through. This means every script that uses the new functionality would automatically update the web interface.

IO Socket sounds good (I think). Also the idea of passing movements from browser to board is awesome.

I also may have changed the colours a little bit :)

![webim](https://user-images.githubusercontent.com/82707095/137298932-7efac49b-a36e-4383-b169-1a440e150420.png)

**Comment by wormstein** [comment]
I couldn't wait till later so I remote home and made a pull on my dev machine just to see with my eyes :)
It is sooo cooll!!! That is such a bice idea, I think it makes even DGT jealous :)))
I'll make the changes later this evening to get the board alive.

**Comment by wormstein** [comment]
We need to have a logo of our project and put it in the web in the top right corner,,, Sonething like DGT Centaur Mods.

**Reply by EdNekebno** [reply]
Yup...I agree. We can also put a logo on the e-paper before it goes to sleep. Then it's best it's 128 pixels wide and resizes to 64 for the web, I think. We should also be careful that it's not too much like DGT's. I can't draw/design though :) Are you creative? or should I fire up paint and see what falls out?

**Reply by wormstein** [reply]
We can start from the logo.bmp that's already in centaur/fonts folder. Resize that and just add Mods word be;low centaur.. Just an idea...  Paint shoud be enough for this :) 
I'm only free later today so feel free to give it a go if you want...

**Comment by wormstein** [comment]
Is it all good with fen.log? I think it needs to chmod to 666 because centaur runs as root. At least that's how it is on my board...
Il'll write this down to remember for our setup...

**Reply by EdNekebno** [reply]
Works fine for me. The numbers i never got used to, but I think it's at 644

**Comment by wormstein** [comment]
I'll see if I can convince RaspiOS to let me run web interface on port 80 without root. Using sudo breaks things...

**Reply by EdNekebno** [reply]
That would be ideal. But :5000 is still good. It's just me that's an idiot and forgets sometimes :)
